hw6/hw6_t02.py

```diff
@@ -0,0 +1,132 @@
+а не просто принт 'You are late'.
+Поднимайте исключение DeadlineError с сообщением 'You are late' вместо print.
+3. Student и Teacher имеют одинаковые по смыслу атрибуты
+(last_name, first_name) - избавиться от дублирования с помощью наследования
+4.
+Teacher
+Атрибут:
+    homework_done - структура с интерфейсом как в словаря, сюда поподают все
+    HomeworkResult после успешного прохождения check_homework
+    (нужно гаранитровать остутствие повторяющихся результатов по каждому
+    заданию), группировать по экземплярам Homework.
+    Общий для всех учителей. Вариант ипользования смотри в блоке if __main__...
+Методы:
+    check_homework - принимает экземпляр HomeworkResult и возвращает True если
+    ответ студента больше 5 символов, так же при успешной проверке добавить в
+    homework_done.
+    Если меньше 5 символов - никуда не добавлять и вернуть False.
+    reset_results - если передать экземпряр Homework - удаляет только
+    результаты этого задания из homework_done, если ничего не передавать,
+    то полностью обнулит homework_done.
+PEP8 соблюдать строго.
+Всем перечисленным выше атрибутам и методам классов сохранить названия.
+К названием остальных переменных, классов и тд. подходить ответственно -
+давать логичные подходящие имена.
+"""
+
+import datetime
+from collections import defaultdict
+from typing import Any, Optional, Union
+
+
+class Homework:
+    def __init__(self, text: str, deadline: int):
+        self.text = text
+        self.deadline = datetime.timedelta(deadline)
+        self.created = datetime.datetime.today()
+
+    def is_active(self) -> bool:
+        """Check if homework is still active."""
+        return datetime.datetime.today() < self.created + self.deadline
+
+
+class DeadlineError(Exception):
+    ...
+
+
+class Person:
+    def __init__(self, last_name: str, first_name: str):
+        self.last_name = last_name
+        self.first_name = first_name
+
+
+class Student(Person):
+    def do_homework(self, homework: Homework, solution: Union[str, int]) -> Any:
+        """Check if not late for homework."""
+        if homework.is_active():
+            return HomeworkResult(self, homework, solution)
+        else:
+            raise DeadlineError("You are late")
+
+
+class HomeworkResult:
+    def __init__(self, author: Student, homework: Homework, solution: Union[str, int]):
+        self.author = author
+        self.solution = str(solution)
+        self.created = datetime.datetime.now()
+        if isinstance(homework, Homework):
+            self.homework = homework
+        else:
+            raise TypeError("Not a Homework object")
+
+    def __eq__(self, other: Any):
+        if isinstance(other, HomeworkResult):
+            return (
+                self.author,
+                self.homework,
+                self.solution == other.author,
+                other.homework,
+                other.solution,
+            )
+        return NotImplemented
```

Comment:
Could you please expand on this code?

Reply:
It returns `NotImplemented` class when comparing the object to the object of another type

Reply:
I have never seen that approach, could you share a link?

Reply:
https://stackoverflow.com/questions/2909106/whats-a-correct-and-good-way-to-implement-hash

Reply:
Do you know that person? Could I trust him?  

Can you give a mo solid link for some trustworthy resource? For example Python documentation.

Reply:
**[SOLID LINK to constants](https://docs.python.org/3/library/constants.html)**
[**SOLID LINK to Implementing the arithmetic operations**](https://docs.python.org/3/library/numbers.html#implementing-the-arithmetic-operations)

Reply:
Keep in mind that using all capital letters is considered shouting.  It may be treated as rude behavior in some cultures.

Also [SOLID](https://en.wikipedia.org/wiki/SOLID) is an abbreviation, using it in all caps may lead to some confusion.